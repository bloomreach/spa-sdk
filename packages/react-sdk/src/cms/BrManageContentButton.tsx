/*
 * Copyright 2019-2025 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useContext } from 'react';
import { ManageContentButton, TYPE_MANAGE_CONTENT_BUTTON } from '@bloomreach/spa-sdk';
import { BrMeta } from '../meta';
import { BrPageContext } from '../page/BrPageContext';

/**
 * The button component that opens for editing a content.
 */
export function BrManageContentButton(props: ManageContentButton): React.ReactElement | null {
  const page = useContext(BrPageContext);

  return page?.isPreview() ? <BrMeta meta={page.getButton(TYPE_MANAGE_CONTENT_BUTTON, props)} /> : null;
}
