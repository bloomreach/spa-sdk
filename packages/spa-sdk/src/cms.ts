/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/// <reference path="./cms.d.ts" />
import { Typed } from 'emittery';
import { Events } from './events';

const GLOBAL_WINDOW = typeof window === 'undefined' ? undefined : window;

export interface CmsOptions {
  /**
   * The window reference for the CMS integration.
   * By default the global window object will be used.
   */
  window?: Window;
}

/**
 * CMS integration layer.
 */
export interface Cms {
  /**
   * Initializes integration with the CMS.
   * @param options The CMS integration options.
   */
  initialize(options: CmsOptions): void;
}

export class Cms implements Cms {
  private api?: CmsApi;
  private postponed: Function[] = [];

  constructor(protected eventBus: Typed<Events>) {}

  private async flush() {
    this.postponed
      .splice(0)
      .forEach(task => task());
  }

  private postpone<T extends (...args: any[]) => any>(task: T) {
    return (...args: Parameters<T>) => {
      if (this.api) {
        return task.apply(this, args);
      }

      this.postponed.push(task.bind(this, ...args));
    };
  }

  initialize({ window = GLOBAL_WINDOW }) {
    if (this.api || !window || window.SPA) {
      return;
    }

    this.eventBus.on('page.ready', this.postpone(this.sync));

    window.SPA = {
      init: this.onInit.bind(this),
      renderComponent: this.onRenderComponent.bind(this),
    };
  }

  protected onInit(api: CmsApi) {
    this.api = api;
    this.flush();
  }

  protected onRenderComponent(id: string, properties: object) {
    this.eventBus.emit('cms.update', { id, properties });
  }

  protected sync() {
    this.api!.sync();
  }
}
