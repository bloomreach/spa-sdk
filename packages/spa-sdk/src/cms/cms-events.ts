/*
 * Copyright 2019-2022 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Typed } from 'emittery';

/**
 * Channel Manager component update event.
 */
export interface CmsUpdateEvent {
  /**
   * Component's id.
   */
  id: string;

  /**
   * Updated component's properties.
   */
  properties: Record<string, unknown>;
}

/**
 * SPA page rendered event.
 */
// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface PageReadyEvent {}

export interface CmsEvents {
  'cms.update': CmsUpdateEvent;
  'page.ready': PageReadyEvent;
}

export type CmsEventBus = Typed<CmsEvents>;
