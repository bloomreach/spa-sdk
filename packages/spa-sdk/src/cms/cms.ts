/*
 * Copyright 2019-2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Typed } from 'emittery';
import { Events, CmsUpdateEvent } from '../events';
import { RpcClient, RpcServer, Procedures } from './rpc';

const GLOBAL_WINDOW = typeof window === 'undefined' ? undefined : window;

export interface CmsOptions {
  /**
   * The window reference for the CMS integration.
   * By default the global window object will be used.
   */
  window?: Window;
}

/**
 * CMS integration layer.
 */
export interface Cms {
  /**
   * Initializes integration with the CMS.
   * @param options The CMS integration options.
   */
  initialize(options: CmsOptions): void;
}

interface CmsProcedures extends Procedures {
  sync(): void;
}

interface CmsEvents {
}

interface SpaProcedures extends Procedures {
}

interface SpaEvents {
}

export class CmsImpl implements Cms {
  constructor(
    protected eventBus: Typed<Events>,
    protected rpcClient: RpcClient<CmsProcedures, CmsEvents>,
    protected rpcServer: RpcServer<SpaProcedures, SpaEvents>,
  ) {
    this.onPageReady = this.onPageReady.bind(this);
  }

  initialize() {
    this.eventBus.on('page.ready', this.onPageReady);
  }

  protected onPageReady() {
    this.rpcClient.call('sync');
  }
}
