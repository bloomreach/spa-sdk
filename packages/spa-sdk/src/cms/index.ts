/*
 * Copyright 2020-2022 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { AsyncContainerModule } from 'inversify';
import { EventBus } from './events';

export const CmsService = Symbol.for('CmsService');
export const EventBusService = Symbol('EventBusService');
export const EventBusServiceProvider = Symbol('EventBusServiceProvider');
export const PostMessageService = Symbol.for('PostMessageService');
export const RpcClientService = Symbol.for('RpcClientService');
export const RpcServerService = Symbol.for('RpcServerService');

export function CmsModule(): AsyncContainerModule {
  return new AsyncContainerModule(async (bind) => {
    const { Typed } = await import('emittery');
    const { CmsImpl } = await import('./cms');
    const { Cms14Impl } = await import('./cms14');
    const { PostMessage } = await import('./post-message');

    bind(EventBusService)
      .toDynamicValue(() => new Typed())
      .inSingletonScope()
      .when(() => typeof window !== 'undefined');
    bind(PostMessageService).to(PostMessage).inSingletonScope();
    bind(RpcClientService).toService(PostMessageService);
    bind(RpcServerService).toService(PostMessageService);
    bind(CmsService).to(CmsImpl).inSingletonScope().whenTargetIsDefault();
    bind(CmsService).to(Cms14Impl).inSingletonScope().whenTargetNamed('cms14');
  });
}

export type EventBusProvider = () => Promise<EventBus | undefined>;
export { CmsOptions, Cms } from './cms';
export { CmsUpdateEvent, EventBus } from './events';
export { PostMessageOptions, PostMessage } from './post-message';
export { RpcClient, RpcServer } from './rpc';
