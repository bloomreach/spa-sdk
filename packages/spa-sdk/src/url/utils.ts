/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

interface Url {
  hash: string;
  origin: string;
  pathname: string;
  search: string;
  searchParams: URLSearchParams;
  path: string;
}

export function buildUrl(url: Partial<Url>) {
  const searchParams = url.searchParams?.toString() ?? '';
  const search = url.search ?? `${searchParams && `?${searchParams}`}`;
  const path = url.path ?? `${url.pathname ?? ''}${search}${url.hash ?? ''}`;

  return `${url.origin ?? ''}${path}`;
}

function isMatchedOrigin(origin: string, baseOrigin: string) {
  return !baseOrigin || !origin || baseOrigin === origin;
}

function isMatchedPathname(pathname: string, basePathname: string) {
  return !basePathname || pathname.startsWith(basePathname);
}

function isMatchedQuery(search: URLSearchParams, baseSearch: URLSearchParams) {
  let match = true;
  baseSearch.forEach((value, key) => {
    match = match && (!value && search.has(key) || search.getAll(key).includes(value));
  });

  return match;
}

export function isMatched(link: string, base = '') {
  const linkUrl = parseUrl(link);
  const baseUrl = parseUrl(base);

  return isMatchedOrigin(linkUrl.origin, baseUrl.origin)
    && isMatchedPathname(linkUrl.pathname, baseUrl.pathname)
    && isMatchedQuery(linkUrl.searchParams, baseUrl.searchParams);
}

export function mergeSearchParams(params: URLSearchParams, ...rest: URLSearchParams[]) {
  const result = new URLSearchParams(params);
  rest.forEach(params => params.forEach((value, key) => result.set(key, value)));

  return result;
}

export function parseUrl(url: string): Url {
  // URL constructor requires either a valid URL or a base URL.
  // Since this function returns a pathname, we can safely pass a fake host to be able to resolve relative URLs.
  const parsedUrl = url ? new URL(url, 'http://example.com') : {} as URL;
  const { hash = '', search = '', searchParams = new URLSearchParams() } = parsedUrl;

  // For links like `//example.com?query#hash` pathname will be `/` so we need to strip query and hash parameters first.
  let origin = url.substring(0, url.length - search.length - hash.length);
  let { pathname = '' } = parsedUrl;
  if (!origin.endsWith(pathname)) {
    pathname = pathname.substring(1);
  }

  origin = origin.substring(0, origin.length - pathname.length);

  return { hash, origin, pathname, search, searchParams, path: `${pathname}${search}${hash}` };
}
