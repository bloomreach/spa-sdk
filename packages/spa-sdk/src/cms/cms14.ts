/*
 * Copyright 2020-2022 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { injectable, inject, optional } from 'inversify';
import { Spa, SpaService } from '../spa';
import { CmsEventBus } from './cms-events';
import { CmsOptions, Cms } from './cms';
import { Logger } from '../logger';
import { CmsEventBusService } from './index';

const GLOBAL_WINDOW = typeof window === 'undefined' ? undefined : window;

interface CmsApi {
  sync(): void;
}

interface SpaApi {
  init(api: CmsApi): void;
  renderComponent(id: string, properties: Record<string, unknown>): void;
}

declare global {
  interface Window {
    SPA?: SpaApi;
  }
}

@injectable()
export class Cms14Impl implements Cms {
  private api?: CmsApi;

  // eslint-disable-next-line @typescript-eslint/ban-types
  private postponed: Function[] = [];

  constructor(
    @inject(SpaService) private spa: Spa,
    @inject(CmsEventBusService) @optional() protected eventBus?: CmsEventBus,
    @inject(Logger) @optional() private logger?: Logger,
  ) {}

  private async flush(): Promise<void> {
    this.postponed.splice(0).forEach((task) => task());
  }

  private postpone<T extends (...args: any[]) => any>(task: T) {
    return (...args: Parameters<T>) => {
      if (this.api) {
        return task.apply(this, args);
      }

      this.postponed.push(task.bind(this, ...args));

      return undefined;
    };
  }

  initialize({ window = GLOBAL_WINDOW }: CmsOptions): void {
    if (this.api || !window || window.SPA) {
      return;
    }

    this.logger?.debug('Initiating a handshake with the Experience Manager.');
    this.eventBus?.on('page.ready', this.postpone(this.sync));

    window.SPA = {
      init: this.onInit.bind(this),
      renderComponent: this.onRenderComponent.bind(this),
    };
  }

  protected onInit(api: CmsApi): void {
    this.logger?.debug('Completed the handshake with the Experience Manager.');

    this.api = api;
    this.flush();
  }

  protected onRenderComponent(id: string, properties: Record<string, unknown>): void {
    this.logger?.debug('Received component rendering request.');
    this.logger?.debug('Component:', id);
    this.logger?.debug('Properties', properties);

    this.spa.onCmsUpdate({ id, properties });
  }

  protected sync(): void {
    this.logger?.debug('Synchronizing the page.');

    this.api!.sync();
  }
}
