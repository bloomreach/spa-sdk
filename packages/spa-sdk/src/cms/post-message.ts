/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { isMatched } from '../url';
import { Events, Message, Procedures, Rpc } from './rpc';

const GLOBAL_WINDOW = typeof window === 'undefined' ? undefined : window;

export interface PostMessageOptions {
  /**
   * CMS origin to verify a message sender.
   */
  origin: string;

  /**
   * The window reference for the CMS integration.
   * By default the global window object will be used.
   */
  window?: Window;
}

export class PostMessage<
  TRemoteProcedures extends Procedures,
  TRemoteEvents extends Events,
  TProcedures extends Procedures,
  TEvents extends Events,
> extends Rpc<TRemoteProcedures, TRemoteEvents, TProcedures, TEvents> {
  private options?: PostMessageOptions;

  constructor() {
    super();

    this.onMessage = this.onMessage.bind(this);
  }

  private get window() {
    return this.options?.window || GLOBAL_WINDOW;
  }

  initialize(options: PostMessageOptions) {
    this.window?.removeEventListener('message', this.onMessage, false);
    this.options = options;
    this.window?.addEventListener('message', this.onMessage, false);
  }

  protected send(message: Message) {
    this.window?.parent?.postMessage(message, this.options!.origin);
  }

  private onMessage(event: MessageEvent) {
    if (!event.data || !this.options || !isMatched(event.origin, this.options.origin)) {
      return;
    }

    this.process(event.data);
  }
}
