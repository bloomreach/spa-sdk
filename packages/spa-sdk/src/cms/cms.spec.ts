/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { mocked } from 'ts-jest/utils';
import { Typed } from 'emittery';
import { Events } from '../events';
import { CmsImpl } from './cms';
import { RpcClient, RpcServer } from './rpc';

describe('CmsImpl', () => {
  let cms: CmsImpl;
  let eventBus: Typed<Events>;
  let rpcClient: RpcClient<any, any>;
  let rpcServer: RpcServer<any, any>;

  beforeEach(() => {
    eventBus = new Typed<Events>();
    rpcClient = {
      call: jest.fn(),
      on: jest.fn(),
      off: jest.fn(),
    };
    rpcServer = {
      register: jest.fn(),
      trigger: jest.fn(),
    };
    cms = new CmsImpl(eventBus, rpcClient, rpcServer);
  });

  describe('initialize', () => {
    const getReadyState = jest.fn();

    beforeEach(() => {
      Object.defineProperty(document, 'readyState', {
        get: getReadyState,
      });
    });

    it('should trigger ready event right away', () => {
      getReadyState.mockReturnValueOnce('interactive');
      cms.initialize({ window });

      expect(rpcServer.trigger).toHaveBeenCalledWith('ready', undefined);
    });

    it('should not trigger ready event on state change if the state is still loading', async () => {
      getReadyState.mockReturnValueOnce('loading');
      getReadyState.mockReturnValueOnce('loading');
      cms.initialize({ window });

      document.dispatchEvent(new ProgressEvent('readystatechange'));
      await new Promise(resolve => setTimeout(resolve, 0));

      expect(rpcServer.trigger).not.toHaveBeenCalled();
    });

    it('should trigger ready event on state change', async () => {
      getReadyState.mockReturnValueOnce('loading');
      getReadyState.mockReturnValueOnce('interactive');
      cms.initialize({ window });

      document.dispatchEvent(new ProgressEvent('readystatechange'));
      await new Promise(resolve => setTimeout(resolve, 0));

      expect(rpcServer.trigger).toHaveBeenCalledWith('ready', undefined);
    });
  });

  describe('onPageReady', () => {
    it('should process postponed events on initialization', async () => {
      cms.initialize({ window });
      await eventBus.emit('page.ready', {});

      expect(rpcClient.call).toHaveBeenCalledWith('sync');
    });
  });

  describe('onUpdate', () => {
    it('should process postponed events on initialization', async () => {
      cms.initialize({ window });

      expect(rpcClient.on).toHaveBeenCalledWith('update', expect.any(Function));

      const [, onUpdate] = mocked(rpcClient.on).mock.calls.pop()!;
      const event = { id: 'id', properties: { a: 'b' } };
      const emitSpy = spyOn(eventBus, 'emit');
      onUpdate(event);

      expect(emitSpy).toHaveBeenCalledWith('cms.update', event);
    });
  });
});
