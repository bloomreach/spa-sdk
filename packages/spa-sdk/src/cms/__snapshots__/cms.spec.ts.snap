// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CmsImpl inject should put a script element 1`] = `
<script
  crossorigin="use-credentials"
  src="url1"
  type="text/javascript"
/>
`;
